### PR TITLE
Miscellaneous fixes

### PR DESCRIPTION
**Pull request by NeilRashbrook**

### CODE REVIEW

File: test/array.test.js [hunk 1]
```diff
@@ -91,9 +94,15 @@ test('Array replace', done => {
 
 test('Array search functions', () => {
   let a = newArray();
+  expect(ArrayColl.isArray(a)).toBe(true);
+  expect(ArrayColl.isArray([])).toBe(true);
+  expect(ArrayColl.isArray({})).toBe(false);
+  expect(ArrayColl.isArrayColl(a)).toBe(true);
+  expect(ArrayColl.isArrayColl([])).toBe(false);
+  expect(ArrayColl.isArrayColl({})).toBe(false);
```
Comment (benbucksch) [inline test/array.test.js:110]: This should be a separate test

File: test/array.test.js [hunk 2]
```diff
@@ -91,9 +94,15 @@ test('Array replace', done => {
...
   expect(a.length).toBe(5);
 
-  expect(a.includes("b")).toBe(true);
```
Comment (benbucksch) [inline test/array.test.js:100]: This line should stay.

File: lib/collection/map.js [hunk 3]
```diff
@@ -64,7 +67,7 @@ export class MapColl extends KeyValueCollection {
    */
   set(key, value) {
     let oldValue = this._map.get(key);
-    if (oldValue == value) {
+    if (oldValue === value) {
```
Comment (benbucksch) [inline lib/collection/map.js:70]: Good catch!  Strict equality is important in this particular case.

File: lib/collection/map.js [hunk 4]
```diff
@@ -38,8 +38,11 @@ export class MapColl extends KeyValueCollection {
   }
 
   clear() {
-    this._notifyRemoved(this.contents);
-    this._map.clear();
+    if (this._map.size) {
+      let items = this.contents;
+      this._map.clear();
+      this._notifyRemoved(items);
+    }
```
Comment (benbucksch) [inline lib/collection/map.js]: Make such things an early return. The normal code path should be straight, not indented.

```suggestion
    if (this.isEmpty) {
      return;
    }
    let items = this.contents;
    this._map.clear();
    this._notifyRemoved(items);
```

File: lib/collection/array.js [hunk 5]
```diff
@@ -235,24 +232,31 @@ export class ArrayColl extends KeyValueCollection {
...
+    this.addAll(items);
+    return this.length;
+  }
+
   shift() {
-    let removedItem = this._array.shift();
-    if (removedItem) {
+    let removedItem;
+    if (this._array.length) {
+      removedItem = this._array.shift();
```
Comment (benbucksch) [inline lib/collection/array.js]: Why this change? The old code was easier to read.

File: lib/collection/array.js [hunk 6]
```diff
@@ -235,24 +232,31 @@ export class ArrayColl extends KeyValueCollection {
...
   join(separator) {
     return this._array.join(separator);
   }
 
   pop() {
-    let removedItem = this._array.pop();
-    if (removedItem) {
+    let removedItem;
+    if (this._array.length) {
+      removedItem = this._array.pop();
```
Comment (benbucksch) [inline lib/collection/array.js]: Why this change? The old code was easier to read.

File: lib/collection/array.js [hunk 7]
```diff
@@ -45,29 +45,25 @@ export class ArrayColl extends KeyValueCollection {
...
    */
   remove(item) {
-    arrayRemove(this._array, item, false);
-    this._notifyRemoved([item]);
+    if (arrayRemove(this._array, item, false)) {
+      this._notifyRemoved([item]);
+    }
   }
 
   _removeWithoutObserver(item) {
```
Comment (benbucksch) [inline lib/collection/array.js:54]: Could you please document what it now returns?

```suggestion
   /** @returns {integer} number of items removed */
  _removeWithoutObserver(item) {
```

File: lib/collection/array.js [hunk 8]
```diff
@@ -158,33 +157,31 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   removeKey(i) {
-    var item = this._array[i];
-    if (item == undefined) {
-      return;
+    if (i in this._array) {
+      this._notifyRemoved(this._array.splice(i, 1));
     }
-    delete this._array[i];
```
Comment (benbucksch) [inline lib/collection/array.js:169]: I think we still need this line, no? Otherwise this function doesn't actually remove anything.

File: lib/collection/array.js [hunk 9]
```diff
@@ -158,33 +157,31 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   removeKey(i) {
-    var item = this._array[i];
-    if (item == undefined) {
-      return;
+    if (i in this._array) {
+      this._notifyRemoved(this._array.splice(i, 1));
```
Comment (benbucksch) [inline lib/collection/array.js]:
```suggestion
      delete this._array[i];
      this._notifyRemoved(this._array.splice(i, 1));
```

File: lib/collection/array.js (hunk 8 above)
Comment (benbucksch) [reply lib/collection/array.js:169]: Please also add a test for this.

File: lib/collection/array.js [hunk 10]
```diff
@@ -158,33 +157,31 @@ export class ArrayColl extends KeyValueCollection {
...
 
   getKeyForValue(value) {
-    for (var i in this._array) {
-      if (this._array[i] == value) {
-        return i;
-      }
-    }
-    return undefined;
+    let key = this._array.indexOf(value);
+    return key < 0 ? undefined : key;
```
Comment (benbucksch) [inline lib/collection/array.js]: Non-error path more prominent:

```suggestion
    return key >= 0 ? key : undefined;
```

File: lib/collection/array.js [hunk 11]
```diff
@@ -158,33 +157,31 @@ export class ArrayColl extends KeyValueCollection {
...
 
-    // Ensure same order and duplicity
-    this._array.length = 0;
     if (coll instanceof Collection) {
-      coll = coll.contents;
+      this._array = coll.contents;
+    } else {
+      this._array = coll.slice();
+    }
+    let addedItems = this._array.filter(item => !arrayRemove(removedItems, item));
```
Comment (benbucksch) [inline lib/collection/array.js:195]: Smart, but hard to read code.
Side effect: changing removedItems while calculating addedItems.

File: lib/collection/array.js (hunk 5 above)
Comment (benbucksch) [inline lib/collection/array.js]:
```suggestion
    let removedItem = this._array.shift();
    if (removedItem) {
```

File: lib/collection/array.js (hunk 6 above)
Comment (benbucksch) [inline lib/collection/array.js]:
```suggestion
    let removedItem = this._array.pop();
    if (removedItem) {
```

File: lib/collection/set.js [hunk 12]
```diff
@@ -26,9 +27,6 @@ export class SetColl extends Collection {
   }
 
   _addWithoutObserver(item) {
-    if (!item && item !== 0) {
-      throw "null objects are not allowed";
-    }
```
Comment (benbucksch) [inline lib/collection/set.js:31]: Why did you remove this? This was deliberate.

I understand that JS Set technically allows this, but I don't want to allow in a Set. (It's OK in array and map, there are good usecases there.)

File: lib/collection/set.js [hunk 13]
```diff
@@ -26,9 +27,6 @@ export class SetColl extends Collection {
   }
 
   _addWithoutObserver(item) {
```
Comment (benbucksch) [inline lib/collection/set.js:28]:
```suggestion
  _addWithoutObserver(item) {
    if (item == null) {
      throw new Error("null/undefined objects are not allowed");
    }
```

File: lib/collection/array.js [hunk 14]
```diff
@@ -45,29 +45,25 @@ export class ArrayColl extends KeyValueCollection {
    * to remove them all.
    */
   remove(item) {
-    arrayRemove(this._array, item, false);
-    this._notifyRemoved([item]);
+    if (arrayRemove(this._array, item, false)) {
```
Comment (benbucksch) [inline lib/collection/array.js]:
```suggestion
    if (this._removeWithoutObserver(item)) {
```

File: lib/collection/set.js [hunk 15]
```diff
@@ -41,31 +39,35 @@ export class SetColl extends Collection {
...
 
   remove(item) {
-    this._removeWithoutObserver(item);
-    this._notifyRemoved([item]);
+    if (this._removeWithoutObserver(item)) {
+      this._notifyRemoved([item]);
+    }
   }
 
   _removeWithoutObserver(item) {
```
Comment (benbucksch) [inline lib/collection/set.js:50]: Document the new return value
```suggestion
  /** @returns {boolean} whether the value was in the set and remove, or not. */
  _removeWithoutObserver(item) {
```

File: lib/collection/set.js [hunk 16]
```diff
@@ -41,31 +39,35 @@ export class SetColl extends Collection {
...
-    this._set.delete(item);
+    return this._set.delete(item);
   }
 
   addAll(items) {
-    for (let item of items) {
-      this._addWithoutObserver(item);
+    items = items.filter(item => this._addWithoutObserver(item));
```
Comment (benbucksch) [inline lib/collection/set.js]: Are you aware that `items` can be a `Collection` object? And that `coll.filter()` will create a dynamic filter object, with observers and all? We should not use that for primitive operations.

Please use the fastest, most light-weight operations for basic operations. (Which are still readable well in the code.)

Function calls are OK (e.g. prefer `this.hasItems` instead of `this._set.size != 0`), and are encouraged for readability and reduction of code duplicaltion, but you need to know what they do.

**IMPORTANT**: Same for all the other functions (which you changed with the purpose of reducing the notifications to the actually removed values).

File: lib/collection/set.js [hunk 17]
```diff
@@ -41,31 +39,35 @@ export class SetColl extends Collection {
...
   }
 
   clear() {
-    this._notifyRemoved(this.contents);
-    this._set.clear();
+    if (this._set.size) {
+      let items = this.contents;
+      this._set.clear();
+      this._notifyRemoved(items);
+    }
```
Comment (benbucksch) [inline lib/collection/set.js]: As above: Please use early exits for errors and edge cases.

```suggestion
    if (this.isEmpty) {
      return;
    }
    let items = this.contents;
    this._set.clear();
    this._notifyRemoved(items);
```

File: lib/operator/subtract.js [hunk 18]
```diff
@@ -87,12 +83,8 @@ class SubtractBaseObserver { // implements CollectionObserver
...
 
   removed(items, coll) {
-    for (let item of items) {
-      if (this.resultColl._collSubtract.contains(item)) {
-        return;
-      }
-      this.resultColl.removeEach(item);
-    }
+    var removeItems = items.filter(item => !this.resultColl._collSubtract.contains(item));
+    this.resultColl.removeAll(removeItems);
```
Comment (benbucksch) [inline lib/operator/subtract.js]: That change introduces a bug. If the base collection contains the removed value multiple times, and the substract coll contains contains it only once, now `removeAll()` removes only a single instance, whereas `removeEach()` removed all of them. The old behavior was correct, the new one leaves substracted values in the output.

Please add a test for this.

Ironically, the old implenentation of `added()` has exactly the same bug, and your new implementation of `added()` fixes it. Funny.

Comment (benbucksch) [reply lib/operator/subtract.js]: I do not understand the goal of why you change the implemenation of all these base operations.

File: lib/operator/subtract.js [hunk 19]
```diff
@@ -102,7 +94,9 @@ class SubtractSubstractedObserver { // implements CollectionObserver
   }
 
   added(items) {
-    this.resultColl.removeAll(items);
+    for (let item of items) {
+      this.resultColl.removeEach(item);
+    }
```
Comment (benbucksch) [inline lib/operator/subtract.js:98]: heh! Here, the old code had the bug described in `removed()`, and your new code fixes it. Good change! :-)

File: test/set.test.js [hunk 20]
```diff
@@ -61,7 +68,7 @@ test('Set forEach', () => {
 test('Set iterators', () => {
   let a = newSet();
   let values = a.values();
-  expect(values.next().value).toBe(a.first);
+  expect(values.first).toBe(a.first);
```
Comment (benbucksch) [inline test/set.test.js]: This change doesn't make sense, because the whole purpose of this test is to check the iterator `.next()`.

Please revert this change. Same below.

File: test/array.test.js [hunk 21]
```diff
@@ -300,10 +309,9 @@ test('Array flat', () => {
 
 test('Array flatMap', () => {
   let a = newArray();
-  a.set(3, newArray().contents);
```
Comment (benbucksch) [inline test/array.test.js:310]: Please leave this line in. The whole purpose of the test is to test nested arrays.

File: test/array.test.js [hunk 22]
```diff
@@ -55,7 +58,7 @@ test('Array shift, unshift', () => {
   let a = newArray();
   let before = a.contents;
 
-  a.unshift(a.shift());
+  a.unshift(a.shift(), a.shift());
```
Comment (benbucksch) [inline test/array.test.js:65]: Please keep a single-param version of the function call in the test. You can add a multi-param version, but the most simple version of the usage should also be covered in the test.

File: test/array.test.js [hunk 23]
```diff
@@ -1,37 +1,37 @@
...
-  return a;
+  return ArrayColl.from(["a", "b", "c", "c", "d"]);
 }
 
 test('Array add, remove', () => {
   let a = newArray();
   expect(a.length).toBe(5);
 
-  a.push("d"); // alias for add()
+  a.push("d", "d"); // alias for addAll([...])
```
Comment (benbucksch) [inline test/array.test.js:11]: Please keep a single-param version of the function call in the test. You can add a multi-param version, but the most simple version of the usage should also be covered in the test.

**Review by benbucksch — CHANGES_REQUESTED** [review]
Overall:
* This fixes small 2-3 bugs
* It reduces the number of observer notification to the actually added/remove items. This is good (but not critical).
* It introduces a few small bugs (see comments)
* It removes a few needed tests (see comments in test code)
* It rewrites some of the base functionality. This comes with risks of regressions.
* When rewriting, pay attention that functions that accept an array also accept a collection. In this case, you are actually calling the operators here, creating new collections, hook up observers etc. We shouldn't do that for the implementation of the most basic operations. Please see the note in one of the review comments.

File: lib/collection/array.js [hunk 24]
```diff
@@ -158,33 +158,32 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   removeKey(i) {
-    var item = this._array[i];
-    if (item == undefined) {
-      return;
+    if (i in this._array) {
+      delete this._array[i];
+      this._notifyRemoved(this._array.splice(i, 1));
     }
```
Comment (benbucksch) [inline lib/collection/array.js]:
```suggestion
    if (!(i in this._array)) {
      return;
    }
    this._notifyRemoved(this._array.splice(i, 1));
```

File: lib/collection/array.js (hunk 8 above)
Comment (benbucksch) [reply lib/collection/array.js:169]: Oh! This was a side effect again, from the `_notifyRemoved()` line! I looked at these lines 3 time, and didn't see it, simply because I didn't expect that.

Please do not do that. No side effects, please. A line that calls `notifyRemoved()` should only notity.
Could you please write this more clearly? E.g. as 2 lines, with a temp variable? E.g.
```
let removedItem = this._array.splice(i, 1);
if (removedItem) {
  this._notifyRemoved(removedItem);
}
```

**Review by benbucksch — CHANGES_REQUESTED** [review]

File: test/array.test.js (hunk 21 above)
Comment (NeilRashbrook) [reply test/array.test.js:310]: That's not how `flatMap` works; it flattens arrays in the result of `map` (to depth 1).

File: test/array.test.js (hunk 2 above)
Comment (NeilRashbrook) [reply test/array.test.js:100]: There is already a test for `includes`; I was just adding one for `contains` as well.

File: lib/collection/array.js (hunk 5 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: Because it fails when the removed item is any of `false`, `0`, `""` or the other falsy values.

File: lib/collection/array.js (hunk 6 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: Ditto the `shift()` comment.

File: lib/collection/array.js (hunk 9 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: This suggestion is definitely wrong.

File: lib/collection/array.js (hunk 5 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: This change re-introduces the bug.

File: lib/collection/array.js (hunk 6 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: This change re-introduces the bug.

File: lib/collection/set.js (hunk 12 above)
Comment (NeilRashbrook) [reply lib/collection/set.js:31]: You're also disallowing all other falsy values other than the `0` that you explicitly check for. If you really want to disallow `null`, you should check for `null`.

Comment (NeilRashbrook) [reply lib/collection/set.js:31]: (And a comment to indicate that this is different behaviour to array and map wouldn't go amiss, of course.)

File: lib/collection/set.js (hunk 13 above)
Comment (NeilRashbrook) [reply lib/collection/set.js:28]: ... so this suggestion is even worse than the original code.

File: lib/collection/set.js (hunk 16 above)
Comment (NeilRashbrook) [reply lib/collection/set.js]: Actually I do remember that biting me at some point, although I forget where now.

File: lib/collection/set.js (hunk 12 above)
Comment (benbucksch) [reply lib/collection/set.js:31]: Agreed on both points. Only null and undefined should be forbidden. Disallowing `false` and `0` was accidental, indeed. I thought I had fixed that, but maybe I didn't. Would you like to fix it? And add the documentation as JSDoc?

File: lib/collection/set.js (hunk 16 above)
Comment (NeilRashbrook) [reply lib/collection/set.js]: I remember now: I was getting collections passed to notifications, which really messes things up when you're trying to do remote notifications.

File: lib/operator/subtract.js (hunk 18 above)
Comment (NeilRashbrook) [reply lib/operator/subtract.js]: No, this is for when you remove items from the base collection; you only want to remove them once from the result collection in this case. (Although, they should not be in the base collection in the first place, so the filter is unnecessary.)

Comment (NeilRashbrook) [reply lib/operator/subtract.js]: ...they should not be in the *result* collection...

File: test/array.test.js (hunk 2 above)
Comment (benbucksch) [reply test/array.test.js:100]: OK, but doesn't hurt to keep this here as well, given that they do the same thing.

File: test/set.test.js (hunk 20 above)
Comment (NeilRashbrook) [reply test/set.test.js]: I was confused as to why the tests were different...

File: test/array.test.js (hunk 21 above)
Comment (benbucksch) [reply test/array.test.js:310]: No matter what the expected result is, we should test an array as array item here.

File: lib/collection/array.js [hunk 25]
```diff
@@ -240,9 +242,8 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   pop() {
-    let removedItem;
-    if (this._array.length) {
-      removedItem = this._array.pop();
+    let removedItem = this._array.pop();
+    if (removedItem) {
```
Comment (NeilRashbrook) [inline lib/collection/array.js]: This reintoduces the bug I was trying to fix.

File: lib/collection/array.js [hunk 26]
```diff
@@ -254,9 +255,8 @@ export class ArrayColl extends KeyValueCollection {
   }
 
   shift() {
-    let removedItem;
-    if (this._array.length) {
-      removedItem = this._array.shift();
+    let removedItem = this._array.shift();
+    if (removedItem) {
```
Comment (NeilRashbrook) [inline lib/collection/array.js]: This reintoduces the bug I was trying to fix.

File: lib/collection/array.js (hunk 5 above)
Comment (benbucksch) [reply lib/collection/array.js]: Duh, right.

Could you at least make an early exit, please?

File: lib/collection/set.js [hunk 27]
```diff
@@ -27,6 +27,9 @@ export class SetColl extends Collection {
   }
 
   _addWithoutObserver(item) {
+    if (!item) {
+      throw new Error("null objects are not allowed");
+    }
```
Comment (NeilRashbrook) [inline lib/collection/set.js]:
```suggestion
    if (item == null) { // null is specifically disallowed in a SetColl
      throw new Error("null objects are not allowed");
    }
```

File: lib/collection/array.js (hunk 25 above)
Comment (NeilRashbrook) [inline lib/collection/array.js]:
```suggestion
    let removedItem;
    if (this._array.length) {
      removedItem = this._array.pop();
```

File: lib/collection/array.js (hunk 26 above)
Comment (NeilRashbrook) [inline lib/collection/array.js]:
```suggestion
    let removedItem;
    if (this._array.length) {
      removedItem = this._array.shift();
```

File: lib/operator/subtract.js (hunk 18 above)
Comment (benbucksch) [reply lib/operator/subtract.js]: Oh, I see. I didn't have enough context in the patch to notice that this is the base coll.

Good point on the subtract coll being irrelevant.

File: lib/collection/set.js (hunk 13 above)
Comment (benbucksch) [reply lib/collection/set.js:28]: Yes, forgot about `0`. Could you fix that, please?

Comment (benbucksch) [reply lib/collection/set.js:28]: I've adapted my suggestion. Does that make more sense?

File: test/set.test.js (hunk 20 above)
Comment (benbucksch) [reply test/set.test.js]: That is deliberate. It tests that the results from various methods are consistent and as expected. If you test the function against itself, it will return the same wrong result twice and pass (false positive).

In any case, the `.next().value` is what this test is really testing, so this needs to stay.

**Review by benbucksch — CHANGES_REQUESTED** [review]

File: lib/collection/array.js (hunk 25 above)
Comment (benbucksch) [reply lib/collection/array.js]: Yes, this should be `if (removedItem != null)` or better yet
```
if (removedItem == null) {
  return;
}
```
or something like that (I cannot see the context anymore)

File: lib/collection/array.js (hunk 26 above)
Comment (benbucksch) [reply lib/collection/array.js]: yes, sorry. See above.

File: lib/collection/set.js (hunk 27 above)
Comment (benbucksch) [reply lib/collection/set.js]: Yes, exactly. Just add this as JSDoc for the function signature, please, so that callers also know.

File: lib/collection/array.js (hunk 25 above)
Comment (benbucksch) [reply lib/collection/array.js]: Use early exit, please.

Comment (benbucksch) [reply lib/collection/array.js]:
```
if (!this._array.length) {
  return;
}
```

File: lib/collection/array.js (hunk 26 above)
Comment (benbucksch) [reply lib/collection/array.js]: Use early exit, please.

```
if (!this._array.length) {
  return;
}
```

File: test/set.test.js (hunk 20 above)
Comment (NeilRashbrook) [reply test/set.test.js]: ... and indeed why the SetColl returns an iterator while the other collections return a collection...

Comment (benbucksch) [reply test/set.test.js]: That's probably a bug in the other collections. The intention is to match the result of the native Array/Set/Map here. I.e.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/values
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/values
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/values

File: lib/operator/subtract.js (hunk 18 above)
Comment (NeilRashbrook) [reply lib/operator/subtract.js]:
```
@@ -87,12 +83,8 @@ class SubtractBaseObserver { // implements CollectionObserver
```
should have been the giveaway.

Would you like me to remove the subtract coll check?

File: lib/collection/set.js (hunk 13 above)
Comment (NeilRashbrook) [reply lib/collection/set.js:28]: Yes, thanks,

File: test/set.test.js (hunk 20 above)
Comment (NeilRashbrook) [reply test/set.test.js]: The methods weren't consistent across collections, which is why I made the original change. I'm not testing the function against itself (ironically your version in `test/array.test.js` which I coped from comes closest to testing the function against itself, as the only difference is that the array is a copy).

File: test/array.test.js (hunk 21 above)
Comment (NeilRashbrook) [reply test/array.test.js:310]: I've added the array back in, but I've changed the callback function (again) because it doesn't make sense for it never to return an array, as otherwise there would be nothing to flatten.

File: lib/collection/set.js (hunk 16 above)
Comment (benbucksch) [reply lib/collection/set.js]: Hey Neil, once you fixed this everywhere (!), could you please make a comment here?

File: lib/operator/subtract.js (hunk 18 above)
Comment (benbucksch) [reply lib/operator/subtract.js]: > Would you like me to remove the subtract coll check?

Yes, please. Thanks for noticing this.

File: lib/collection/set.js [hunk 28]
```diff
@@ -41,31 +42,43 @@ export class SetColl extends Collection {
...
 
   addAll(items) {
-    for (let item of items) {
-      this._addWithoutObserver(item);
+    if (items instanceof Collection) {
+      items = items.contents;
+    }
+    items = items.filter(item => this._addWithoutObserver(item));
+    if (items.length) {
+      this._notifyAdded(items);
```
Comment (benbucksch) [inline lib/collection/set.js]:
```suggestion
    let itemsChanged = items.filter(item => this._addWithoutObserver(item));
    if (itemsChanged.length) {
      this._notifyAdded(itemsChanged);
```

File: lib/collection/set.js [hunk 29]
```diff
@@ -41,31 +42,43 @@ export class SetColl extends Collection {
...
 
   removeAll(items) {
-    for (let item of items) {
-      this._removeWithoutObserver(item);
+    if (items instanceof Collection) {
+      items = items.contents;
+    }
+    items = items.filter(item => this._removeWithoutObserver(item));
+    if (items.length) {
+      this._notifyRemoved(items);
```
Comment (benbucksch) [inline lib/collection/set.js]:
```suggestion
    let itemsChanged = items.filter(item => this._removeWithoutObserver(item));
    if (itemsChanged.length) {
      this._notifyRemoved(itemsChanged);
```

File: lib/collection/set.js (hunk 16 above)
Comment (benbucksch) [reply lib/collection/set.js]: A `for` loop creates only one array.

(same for remove below)

**Review by benbucksch — CHANGES_REQUESTED** [review]

File: lib/collection/set.js (hunk 28 above)
Comment (benbucksch) [reply lib/collection/set.js]: Can we use `for .. of`? (same below)

File: lib/collection/array.js (hunk 25 above)
Comment (NeilRashbrook) [reply lib/collection/array.js]: The reason you can't see the context is that I've switched back to my code whereby the code checks that the array actually has an element that can be popped before trying to pop it rather than hoping that the array never contains a value that can be mistaken for popping from an empty array.

(The above comment is dated Dec 19 although I received the email notification two days ago. And a copy of the comment appears lower down as a child of "benbucksch requested changes 2 days ago" [as of time of writing]. How do you make a comment dated Dec 19 on a change requested 2 days ago? Has it just been waiting in a GitHub review all this time? That's really confusing...)

File: lib/collection/set.js (hunk 16 above)
Comment (NeilRashbrook) [reply lib/collection/set.js]: I've switched all three places over to using a `for` loop in the latest version.

File: lib/operator/subtract.js (hunk 18 above)
Comment (NeilRashbrook) [reply lib/operator/subtract.js]: I've removed the useless check in the latest version.

File: test/array.test.js (hunk 23 above)
Comment (NeilRashbrook) [reply test/array.test.js:11]: There's already one specifically in the `push`/`pop` test.

File: test/array.test.js (hunk 22 above)
Comment (NeilRashbrook) [reply test/array.test.js:65]: (This had already been restored in the previous version.)

File: test/array.test.js (hunk 1 above)
Comment (NeilRashbrook) [reply test/array.test.js:110]: (This had already been separated in the previous version.)

File: test/set.test.js (hunk 20 above)
Comment (NeilRashbrook) [reply test/set.test.js]: I've now changed `ArrayColl` and `MapColl`'s iterators to match `SetColl` in the latest version.

File: lib/collection/array.js (hunk 25 above)
Comment (benbucksch) [reply lib/collection/array.js]: I don't know how that happened. I didn't write this comment recently.

**Review by benbucksch — APPROVED** [review]
